packetbeat/sniffer/sniffer.go

```diff
@@ -134,15 +140,124 @@ func New(testMode bool, filter string, factory WorkerFactory, interfaces config.
 // Run opens the sniffing device and processes packets being read from that device.
 // Worker instances are instantiated as needed.
 func (s *Sniffer) Run() error {
-	handle, err := s.open()
+	var defaultRoute chan string
+	if s.config.PollDefaultRoute > 0 && strings.HasPrefix(s.config.Device, "default_route") {
+		s.done = make(chan struct{})
+		defaultRoute = make(chan string)
+		go s.pollDefaultRoute(defaultRoute)
+	}
+	if defaultRoute == nil {
+		return s.sniffStatic(s.device)
+	}
+	return s.sniffDynamic(defaultRoute)
+}
+
+// pollDefaultRoute repeatedly polls the default route's device at intervals
+// specified in config.PollDefaultRoute. The poller is terminated by closing
+// done and the device chan can be read for changes in the default route.
+// Changes in default route will put the Sniffer into the inactive state to
+// trigger a new sniffer connection. Termination of the sniffer is not under
+// the control of the poller.
+func (s *Sniffer) pollDefaultRoute(device chan string) {
+	go func() {
+		logp.Info("starting default route poller")
+
+		// Prime the channel.
+		current := s.device
+		device <- current
+
+		tick := time.NewTicker(s.config.PollDefaultRoute)
+		for {
+			select {
+			case <-tick.C:
+				logp.Debug("sniffer", "polling default route")
+				dev, err := resolveDeviceName(s.config.Device)
+				if err != nil {
+					logp.Warn("sniffer failed to poll default route device: %v", err)
+					continue
+				}
+				if dev != current {
+					logp.Info("sniffer changing default route device: %s -> %s", current, dev)
+					current = dev
+					s.state.Store(snifferInactive) // Mark current device as stale. ¯\_(ツ)_/¯
+					device <- current              // Pass the new device name.
+				}
+			case <-s.done:
+				logp.Info("closing default route poller")
+				close(device)
+				tick.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// sniffStatic performs the sniffing work on a single static interface.
+func (s *Sniffer) sniffStatic(device string) error {
+	handle, err := s.open(device)
 	if err != nil {
 		return fmt.Errorf("failed to start sniffer: %w", err)
 	}
 	defer handle.Close()
 
+	worker, err := s.factory(handle.LinkType())
+	if err != nil {
+		return err
+	}
+
+	return s.sniffHandle(handle, worker)
+}
+
+// sniffDynamic performs sniffing work on a stream of dynamic interfaces from
+// defaultRoute workers are retained between successive interfaces if they are
+// the same link type.
+func (s *Sniffer) sniffDynamic(defaultRoute <-chan string) error {
+	var (
+		last   layers.LinkType
+		worker Worker
+	)
+	for device := range defaultRoute {
+		var err error
+		last, worker, err = s.sniffOneDynamic(device, last, worker)
+		if err != nil {
+			return err
```

Comment:
I'm wondering about an edge case that we probably want to handle. If the default route moves to an unsupported link type temporarily, do we want Packetbeat to continue watching for changes even if it cannot sniff the current interface?

Reply:
Do you think that should have a timeout or just continue forever? I'm leaning towards forever.

Reply:
Yeah, I'm thinking forever too. If I leave home and turn on my VPN, which cannot be monitored due to it being a raw device, I still want packetbeat to start monitoring when the VPN goes off.

side note: I think it would be useful to have a monitoring metric that reports the current interface name.

Reply:
This will work for the default route case, but when there are multiple interfaces this won't work AFAICS. So I think the best thing to do is to publish it as the default route.